### PR TITLE
Add ServiceAccountJwtHeaderCredentials

### PR DESCRIPTION
**Pull request by anthmgoogle**
This change is dependent on the previous pull request, so only the 2nd commit needs separate review. When the first change is in I'll revise this. 

@ejona86 @tbetbetbe can you review?

### CODE REVIEW

**Comment by ejona86** [comment]
@anthmgoogle, except for issues above, LGTM

**Comment by anthmgoogle** [comment]
@ejona86 @louiscryan 

Changes since last time:
- Synced the dependent change to allow the URI to be passed.
- Added the licence header
- Added support for the Audience to be passed either in the constructor or in getRequestMetadata to determine which will work best for gRPC.

This should be ready to merge now if there are no further issues.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2014, Google Inc. All rights reserved.
```
Comment (ejona86) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: 2015?

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  /**
+   * Constructor with full information.
+   *
+   * @param clientId Client ID of the service account from the console. May be null.
+   * @param clientEmail Client email address of the service account from the console.
+   * @param privateKey RSA private key object for the service account.
+   * @param privateKeyId Private key identifier for the service account. May be null.
+   * @param defaultAudience Audience to use if not provided by transport. My be null.
```
Comment (ejona86) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: s/My/May/

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  /**
+   * Factory using PKCS#8 for the private key.
+   *
+   * @param clientId Client ID of the service account from the console. May be null.
+   * @param clientEmail Client email address of the service account from the console.
+   * @param privateKeyPkcs8 RSA private key object for the service account in PKCS#8 format.
+   * @param privateKeyId Private key identifier for the service account. May be null.
+   * @param defaultAudience Audience to use if not provided by transport. My be null.
```
Comment (ejona86) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: s/My/May/

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    if (uri == null) {
+      if (defaultAudience != null) {
+        uri = defaultAudience;
+      } else {
+        throw new IOException("JwtAccess requires Audience uri to be passed in or the " 
+          + "defaultAudience to be specified");
+      }
+    }
+    String assertion = getJwtAccess(uri);
+    Map<String, List<String>> newRequestMetadata = new HashMap<String, List<String>>();
```
Comment (ejona86) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Collections.singletonMap() ?

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    List<String> newAuthorizationHeaders = new ArrayList<String>();
```
Comment (ejona86) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Collections.singletonList() ?

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+import com.google.api.client.json.webtoken.JsonWebSignature;
+import com.google.auth.Credentials;
+import com.google.auth.TestUtils;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.IOException;
+import java.lang.NullPointerException;
```
Comment (ejona86) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java]: Unnecessary import.

**Comment by ejona86** [comment]
@anthmgoogle, LGTM

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 1 above)
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Fixed

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 2 above)
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Fixed.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 3 above)
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Fixed.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java (hunk 6 above)
Comment (anthmgoogle) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java]: Fixed.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 4 above)
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Fixed.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 5 above)
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Done.

**Comment by anthmgoogle** [comment]
Applied all suggestions and repushed.

**Pull request merged**